### PR TITLE
ESQL: Replace AssertionError in mappers

### PR DESCRIPTION
**Pull request by alex-spies**
We want to avoid throwing errors; AssertionErrors are thrown in mappers whenever an operation is not supported for whatever reason. Throw UnsupportedOperationException instead.

### CODE REVIEW

**Comment by alex-spies** [comment]
@elasticsearchmachine run elasticsearch-ci/bwc

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/AbstractConvertFunction.java [hunk 1]
```diff
@@ -43,7 +43,7 @@ protected Supplier<EvalOperator.ExpressionEvaluator> evaluator(Supplier<EvalOper
         DataType sourceType = field().dataType();
         var evaluator = evaluators().get(sourceType);
         if (evaluator == null) {
-            throw new AssertionError("unsupported type [" + sourceType + "]");
+            throw new UnsupportedOperationException("unsupported type [" + sourceType + "]");
```
Comment (bpintea) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/AbstractConvertFunction.java:46]: We have some QL Exceptions, that split in sever and client and `EsqlIllegalArgumentException` might be a good choice for these ones.

Comment (alex-spies) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/AbstractConvertFunction.java:46]: This looks more appropriate to me, as it seems to be intended for server-side errors that should've been handled during validation.

@luigidellaquila suggested `UnsupportedOperationException` [here](https://github.com/elastic/elasticsearch/pull/98352#pullrequestreview-1580262191), but if we're all happy with `EsqlIllegalArgumentException` I'm going to go by your suggestion instead.

**Review by bpintea — APPROVED** [review]
LGTM

**Review by costin — APPROVED** [review]
Use EsqlIllegalArgumentException - it's a specific error to ESQL so it can be properly caught and is REST friendly.
UOE and IAE are generic exception, they are thrown even by the JDK and thus cannot be easily be traced.
The compute module uses them to avoid depending on QL however inside *QL we should throw our own exceptions.

**Comment by alex-spies** [comment]
> Use EsqlIllegalArgumentException - it's a specific error to ESQL so it can be properly caught and is REST friendly. UOE and IAE are generic exception, they are thrown even by the JDK and thus cannot be easily be traced. The compute module uses them to avoid depending on QL however inside *QL we should throw our own exceptions.

Alright, I'll use `EsqlIllegalArgumentException` then! For this to fully make sense, I think I'll add another commit on top that replaces some `UnsupportedOperationException`s so we consistently use `EsqlIllegalArgumentException`.

**Comment by alex-spies** [comment]
Alright, now this PR introduces consistent usage of `EsqlIllegalArgumentException` across the the `esql` package, except for compute and tests.

Since this touches a lot more places now, a re-review would be good @costin or @bpintea . If you think this makes sense, I'll refactor using one or two new helper functions because a lot of time the `EsqlIllegalArgumentException`s (formerly: `UnsupportedOperationException`s) are thrown in a very similar pattern.

**Pull request closed without merging**